fuse_core/include/fuse_core/util.hpp

```diff
@@ -150,6 +151,180 @@ Eigen::Matrix<T, 2, 2, Eigen::RowMajor> rotationMatrix2D(const T angle)
+    jacobian_map << 
+      z[0], -z[1], -z[2], -z[3],
+      z[1],  z[0], -z[3],  z[2],
+      z[2],  z[3],  z[0], -z[1],
+      z[3], -z[2],  z[1],  z[0];
+  }
+}
+
+/**
+ * @brief Compute quaternion to AngleAxis conversion and the function jacobian
+ * 
+ * @param[in] q Pointer to the quaternion array (4x1)
+ * @param[in] angle_axis Pointer to the angle_axis array (3x1) 
+ * @param[in] jacobian Pointer to the jacobian matrix (3x4, optional)
+ */
+static inline void quaternionToAngleAxis(const double * q, double * angle_axis, double * jacobian = nullptr) 
+{
+  ceres::QuaternionToAngleAxis(q, angle_axis);
+  if (jacobian) {
+    Eigen::Map<Eigen::Matrix<double, 3, 4, Eigen::RowMajor>> jacobian_map(jacobian);
+    const double & q0 = q[0];
+    const double & q1 = q[1];
+    const double & q2 = q[2];
+    const double & q3 = q[3];
+    const double q_sum2 = q0 * q0 + q1 * q1 + q2 * q2 + q3 * q3;
+    const double sin_theta2 = q1 * q1 + q2 * q2 + q3 * q3;
+    const double sin_theta = std::sqrt(sin_theta2);
+    const double cos_theta = q0;
+
+    if (std::fpclassify(sin_theta) != FP_ZERO) {
+      const double two_theta = 2.0 * 
+        (cos_theta < 0.0 ? std::atan2(-sin_theta, -cos_theta) : std::atan2(sin_theta, cos_theta));  
+      jacobian_map(0, 0) = -2.0 * q1 / q_sum2;
+      jacobian_map(0, 1) = 
+        two_theta / sin_theta + 
+        (2.0 * q0 * q1 * q1) / (sin_theta2 * q_sum2) - 
+        (q1 * q1 * two_theta) / std::pow(sin_theta2, 1.5);
+      jacobian_map(0, 2) = 
+        (2.0 * q0 * q1 * q2) / (sin_theta2 * q_sum2) - 
+        (q1 * q2 * two_theta) / std::pow(sin_theta2, 1.5);
+      jacobian_map(0, 3) =
+        (2.0 * q0 * q1 * q3) / (sin_theta2 * q_sum2) - 
+        (q1 * q3 * two_theta) / std::pow(sin_theta2, 1.5);
+      
+      jacobian_map(1, 0) = -2.0 * q2 / q_sum2;
+      jacobian_map(1, 1) = 
+        (2.0 * q0 * q1 * q2) / (sin_theta2 * q_sum2) - 
+        (q1 * q2 * two_theta) / std::pow(sin_theta2, 1.5);
+      jacobian_map(1, 2) = 
+        two_theta / sin_theta + 
+        (2.0 * q0 * q2 * q2) / (sin_theta2 * q_sum2) - 
+        (q2 * q2 * two_theta) / std::pow(sin_theta2, 1.5);
+      jacobian_map(1, 3) = 
+        (2.0 * q0 * q2 * q3) / (sin_theta2 * q_sum2) - 
+        (q2 * q3 * two_theta) / std::pow(sin_theta2, 1.5);
+      
+      jacobian_map(2, 0) = -2.0 * q3 / q_sum2;
+      jacobian_map(2, 1) = 
+        (2.0 * q0 * q1 * q3) / (sin_theta2 * q_sum2) - 
+        (q1 * q3 * two_theta) / std::pow(sin_theta2, 1.5);
+      jacobian_map(2, 2) = 
+        (2.0 * q0 * q2 * q3) / (sin_theta2 * q_sum2) - 
+        (q2 * q3 * two_theta) / std::pow(sin_theta2, 1.5);
+      jacobian_map(2, 3) =
+        two_theta / sin_theta + 
+        (2.0 * q0 * q3 * q3) / (sin_theta2 * q_sum2) - 
+        (q3 * q3 * two_theta) / std::pow(sin_theta2, 1.5);
+    } else {
+      jacobian_map.setZero();
+      jacobian_map(1, 1) = 2.0;
+      jacobian_map(2, 2) = 2.0;
+      jacobian_map(3, 3) = 2.0;
```

Comment:
```suggestion
      jacobian_map(0, 0) = 2.0;
      jacobian_map(1, 1) = 2.0;
      jacobian_map(2, 2) = 2.0;
```

I believe (3,3) is out of bounds and this is a typo.

Reply:
Thank you for the findings! I checked that function and I believe that yes, there is an error since the jacobian is a 3x4 matrix so index (3,3) is out of bounds, but I believe that your suggested changes are not entirely correct. 
I implemented the jacobians here by symbolic differentiation of https://github.com/ceres-solver/ceres-solver/blob/c29b5257e23f91d6a47c4db9d57350ed4985ea46/include/ceres/rotation.h#L345. Specifically, this is the case where the quaternion represents a "zero" rotations, so lines: https://github.com/ceres-solver/ceres-solver/blob/c29b5257e23f91d6a47c4db9d57350ed4985ea46/include/ceres/rotation.h#L378-L387.

The jacobian of function output (angle axis) wrt input (quaternion) is then a 3x4 matrix where elements in every line are the derivative of angle axis component wrt all the quaternion components in the order `qw, qx, qy, qz`, resulting in:

```cpp
jacobian = [0 2 0 0
            0 0 2 0
            0 0 0 2]
```

So the changes should be:

```cpp
jacobian_map(0, 1) = 2.0;
jacobian_map(1, 2) = 2.0;
jacobian_map(2, 3) = 2.0;
```
Do you agree?

Reply:
Yes, that makes sense! Good catch.